### PR TITLE
Attach Trace Context when an ANR is detected (ANRv1)

### PR DESCRIPTION
**Pull request by markushi**
## :scroll: Description
This draft PR
* improves the accuracy of detecting ANRs
* Aborts ongoing transactions and links them to ANR events

## :bulb: Motivation and Context
Link transactions and ANRs to get more insights (suspect spans, profiles, etc...)

## :green_heart: How did you test it?
Added unit tests

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [x] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [x] No breaking change or entry added to the changelog.
- [x] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: sentry/src/main/java/io/sentry/ITransaction.java [hunk 1]
```diff
@@ -79,6 +79,20 @@ public interface ITransaction extends ISpan {
...
+  /**
+   * Force finishes the transaction and it's child spans with the specified status. If the
+   * transaction is already finished this is a no-op.
+   *
+   * @param status The status to set the unfinished child spans / transaction to.
+   * @param forceKeep Ensures the transaction is kept even when it e.g. contains no child spans.
+   *     Useful for situations were the transaction and profile provide crucial context (e.g. ANRs)
+   */
+  @ApiStatus.Internal
+  void forceFinish(@NotNull final SpanStatus status, boolean forceKeep);
```
Comment (romtsn) [inline sentry/src/main/java/io/sentry/ITransaction.java:91]: tbqh this `forceFinish` + `forceKeep` made me confused 😅 would it make sense to change the naming a bit, something more explicit like `keepIfIdleChildless` maybe? Or we could invert it to `true` and rename to `dropChildless`. Your call

File: sentry-android-core/src/main/java/io/sentry/android/core/ANRWatchDog.java [hunk 2]
```diff
@@ -100,48 +126,49 @@ public void run() {
...
       }
+      // when list is empty, or there's no element NOT_RESPONDING, we can assume the app is not
+      // blocked
+      return false;
     }
+    return true;
+  }
+
+  @FunctionalInterface
+  interface TimeProvider {
```
Comment (romtsn) [inline sentry-android-core/src/main/java/io/sentry/android/core/ANRWatchDog.java:170]: `l`: we also have `ICurrentDateProvider/AndroidCurrentDateProvider`, would it make sense to reuse it here instead?

File: sentry-android-core/src/main/java/io/sentry/android/core/ANRWatchDog.java [hunk 3]
```diff
@@ -22,60 +22,83 @@ final class ANRWatchDog extends Thread {
...
       long timeoutIntervalMillis,
       boolean reportInDebug,
       @NotNull ANRListener listener,
       @NotNull ILogger logger,
       final @NotNull Context context) {
-    this(timeoutIntervalMillis, reportInDebug, listener, logger, new MainLooperHandler(), context);
+    this(
+        () -> SystemClock.uptimeMillis(),
+        timeoutIntervalMillis,
+        500,
```
Comment (romtsn) [inline sentry-android-core/src/main/java/io/sentry/android/core/ANRWatchDog.java:49]: dunno if posting to main thread every 500ms is fine, but I guess it is?

**Review by romtsn — APPROVED** [review]
LGTM, thanks for this 🚀 as discussed internally let's try to finish the scope tx via hints so we can make it work for uncaught exceptions as well

File: sentry/src/main/java/io/sentry/ITransaction.java (hunk 1 above)
Comment (markushi) [reply sentry/src/main/java/io/sentry/ITransaction.java:91]: yes this is awful! I even had the inverse variant in an early draft but couldn't find a right names for it. `dropChildless` sounds good, maybe `dropIfNoChildren` is even more readable?

Comment (romtsn) [reply sentry/src/main/java/io/sentry/ITransaction.java]: yeah `dropIfNoChildren` sounds good 👍

File: sentry-android-core/src/main/java/io/sentry/android/core/ANRWatchDog.java (hunk 3 above)
Comment (markushi) [reply sentry-android-core/src/main/java/io/sentry/android/core/ANRWatchDog.java:49]: Yes, the task posted is pretty lightweight as well - so this shouldn't have any impact

**Pull request merged**